### PR TITLE
Upgrade Babel parser to support Typescript 5.0

### LINKED ISSUE #768: Breaks when using `satisfies` operator in TypeScript
I think this is due to the used babel version, it has been released in `7.20.0` https://github.com/babel/babel/pull/14211

### PR DESCRIPTION
**Pull request by Karfroth**
This PR upgrades Babel parser to 7.21.3, which supports Typescript 5.0. As it is minor version upgrade, I would expect that there will be no breaking changes.

This will help checking the typescript files with new keywords, such as `satisfies`. Hence, this PR can fix #768

To ensure the new syntex support, this PR also adds a simple function to the index.ts file in the test.

### CODE REVIEW

**Review by danyf90 — APPROVED** [review]

**Comment by jportner** [comment]
I tried using this upgraded dependency (via a yarn dependency resolution) and it seems to have completely broken TypeScript parsing on my project. I get a bunch of "TypeError: p is not iterable" errors as a result.

It seems this was a previously documented issue (https://github.com/depcheck/depcheck/issues/688) and it was [fixed in depcheck 1.4.3](https://github.com/depcheck/depcheck/compare/1.4.2...1.4.3) by pinning @babel/parser to exactly 7.6.14. There's a related discussion here too (https://github.com/depcheck/depcheck/issues/687).

TL;DR unfortunately it seems that upgrading @babel/parser to 7.21.3 re-introduces this problem :(

**Comment by Karfroth** [comment]
Thanks for the comment. I will try check the root cause.

**Comment by Karfroth** [comment]
@jportner I tried to dig more and I believe this PR will not cause the same issue again. because the fix for the root cause is already merged to the main branch with #745

**Comment by jportner** [comment]
> @jportner I tried to dig more and I believe this PR will not cause the same issue again. because the fix for the root cause is already merged to the main branch with #745

Oh I see, that has been merged but not released yet.
I was wondering why your unit test passed!

Thanks for double-checking, and apologies for the false alarm 😄

**Comment by danyf90** [comment]
@rumpl do you think we can merge this and get a new release to fix support for TS5? 🙏

**Comment by danyf90** [comment]
@rumpl / @jportner  do you think we can merge this and get a new release to fix support for TS5? 🙏

**Comment by seansfkelley** [comment]
As a workaround, if you're using Yarn, you can both peg the resolved version of the Babel parser (a la https://github.com/depcheck/depcheck/pull/788#issuecomment-1505676628) and then `yarn patch depcheck` to manually patch in the fix to #688, which is thankfully very easy to apply in the transpiled `.js`.

File: package.json [hunk 1]
```diff
@@ -49,7 +49,7 @@
   "license": "MIT",
   "readmeFilename": "README.md",
   "dependencies": {
-    "@babel/parser": "7.16.4",
+    "@babel/parser": "7.21.3",
```
Comment (znarf) [inline package.json:52]:
```suggestion
    "@babel/parser": "7.22.5",
```

Maybe update to the latest.

File: package.json [hunk 2]
```diff
@@ -49,7 +49,7 @@
...
     "@babel/traverse": "^7.12.5",
```
Comment (znarf) [inline package.json:53]:
```suggestion
    "@babel/traverse": "7.22.5",
```

Maybe update that one too for consistency.

File: package.json (hunk 1 above)
Comment (Karfroth) [reply package.json]: Thank you. I did this manually to update lock file as well.

File: package.json (hunk 2 above)
Comment (Karfroth) [reply package.json]: Same as `@babel/parser`, I did this manually to update lock file.

**Review by znarf — APPROVED** [review]
Good for me.

**Comment by danyf90** [comment]
@znarf @jportner is there any blocker for merging and releasing this? 🙏

**Comment by rumpl** [comment]
> @znarf @jportner is there any blocker for merging and releasing this? 🙏

None, I’ll do it this weekend when I get back from my PTO

**Comment by danyf90** [comment]
Thanks for merging this @rumpl 🙏 
Do you think we can do a release containing this change? 🙏

**Comment by jtbandes** [comment]
I'm also waiting for this change to be released 🎉  thank you!

**Comment by PeKne** [comment]
Thank you for fixing this. Looking forward to the release 🙏.

**Comment by pkishorez** [comment]
@rumpl thanks a lot for your efforts.
I could see some changes merged to main, yet no release created since last year!
Is there any issue regarding the same?
